File: 0070-climbing-stairs/0070-climbing-stairs.cpp
class Solution {
public:
    int climbStairsHelper(int n, vector<int>& memo) {
        // Base cases
        if (n <= 1) return 1;
        
        // If already computed, return the stored result
        if (memo[n] != -1) return memo[n];
        
        // Compute the result and store it in memo
        memo[n] = climbStairsHelper(n - 1, memo) + climbStairsHelper(n - 2, memo);
        return memo[n];
    }
    int climbStairs(int n) {
        // if(n<=1) return 1;
        // vector<int> dp(n+1,0);
        // dp[0]=1;
        // dp[1]=1;
        // for(int i=2;i<=n;i++){
        //     dp[i]= dp[i-1]+dp[i-2];
        // }
        // return dp[n];
        vector<int> memo(n + 1, -1);
        return climbStairsHelper(n, memo);
    }
};

File: 0070-climbing-stairs/0070-climbing-stairs.java
class Solution {
    // Bottom-up approach (Tabulation)
    public int climbStairs(int n) {
        if (n <= 1) return 1;
        
        int[] dp = new int[n + 1];
        dp[0] = 1;
        dp[1] = 1;
        
        for (int i = 2; i <= n; i++) {
            dp[i] = dp[i-1] + dp[i-2];
        }
        
        return dp[n];
    }

    // Top-down approach (Memoization)
    public int climbStairsMemo(int n) {
        int[] memo = new int[n + 1];
        return climbStairsHelper(n, memo);
    }
    
    private int climbStairsHelper(int n, int[] memo) {
        if (n <= 1) return 1;
        if (memo[n] != 0) return memo[n];
        
        memo[n] = climbStairsHelper(n-1, memo) + climbStairsHelper(n-2, memo);
        return memo[n];
    }
}